pkg/standalone-utils/test/foundry/EclpLPOracle.t.sol

```diff
@@ -0,0 +1,648 @@
+            GyroECLPPool(address(pool)),
+            oraclePricesScaled18[0].divDown(oraclePricesScaled18[1])
+        );
+        vm.revertTo(snapshotId);
+
+        return
+            marketPriceBalances[0].mulDown(oraclePricesScaled18[0]) +
+            marketPriceBalances[1].mulDown(oraclePricesScaled18[1]);
+    }
+
+    function _findBalancesForPrices(
+        GyroECLPPool pool,
+        uint256 oraclePrice
+    ) private returns (uint256[] memory marketPriceBalances) {
+        IERC20[] memory tokens;
+        uint256[] memory balancesRaw;
+        (tokens, , balancesRaw, marketPriceBalances) = vault.getPoolTokenInfo(address(pool));
+
+        (IGyroECLPPool.EclpParams memory eclpParams, IGyroECLPPool.DerivedEclpParams memory derivedEclpParams) = pool
+            .getECLPParams();
+
+        uint256 limitTokenABalance = uint256(0);
+        uint256 limitTokenBBalance = uint256(0);
+
+        for (uint256 i = 0; i < 255; i++) {
+            (int256 a, int256 b) = _computeOffsetFromBalances(marketPriceBalances, eclpParams, derivedEclpParams);
+            uint256 price = _computePrice(marketPriceBalances, eclpParams, a, b);
+
+            if (
+                (price > oraclePrice && (price - oraclePrice).divDown(oraclePrice) < 1e6) ||
+                (price < oraclePrice && (oraclePrice - price).divDown(oraclePrice) < 1e6)
+            ) {
+                return marketPriceBalances;
+            }
+
+            if (price > oraclePrice) {
+                // Overpriced (more B than A) => Swap A for B, exact out
+                uint256 exactAmountOut = (balancesRaw[1] - limitTokenBBalance) / 2;
+                limitTokenABalance = balancesRaw[0];
+
+                vm.prank(lp);
+                router.swapSingleTokenExactOut(
+                    address(pool),
+                    tokens[0],
+                    tokens[1],
+                    exactAmountOut,
+                    MAX_UINT256,
+                    MAX_UINT256,
+                    false,
+                    bytes("")
+                );
+
+                (, , balancesRaw, marketPriceBalances) = vault.getPoolTokenInfo(address(pool));
+            } else {
+                // Underpriced (more A than B) => Swap B for A, exact out
+                uint256 exactAmountOut = (balancesRaw[0] - limitTokenABalance) / 2;
+                limitTokenBBalance = balancesRaw[1];
+
+                vm.prank(lp);
+                router.swapSingleTokenExactOut(
+                    address(pool),
+                    tokens[1],
+                    tokens[0],
+                    exactAmountOut,
+                    MAX_UINT256,
+                    MAX_UINT256,
+                    false,
+                    bytes("")
+                );
+
+                (, , balancesRaw, marketPriceBalances) = vault.getPoolTokenInfo(address(pool));
+            }
+        }
+
+        revert("Could not find market balances");
+    }
+
+    function _computePrice(
```

Comment:
This is part of the surge hook right? Maybe we can put that in a library and borrow it from there? (can be done in a follow-up PR).